12272020/root/opt/joy2key/RetroPie-Setup/scriptmodules/supplementary/runcommand/joy2key.py

```diff
@@ -0,0 +1,296 @@
+# https://raw.githubusercontent.com/RetroPie/RetroPie-Setup/master/LICENSE.md
+import os, sys, struct, time, fcntl, termios, signal
+import curses, errno, re
+from pyudev import Context
+
+
+#    struct js_event {
+#        __u32 time;     /* event timestamp in milliseconds */
+#        __s16 value;    /* value */
+#        __u8 type;      /* event type */
+#        __u8 number;    /* axis/button number */
+#    };
+
+JS_MIN = -32768
+JS_MAX = 32768
+JS_REP = 0.20
+
+JS_THRESH = 0.75
+
+JS_EVENT_BUTTON = 0x01
+JS_EVENT_AXIS = 0x02
+JS_EVENT_INIT = 0x80
+
+CONFIG_DIR = '/home/ark/.config/'
+RETROARCH_CFG = CONFIG_DIR + 'retroarch/retroarch.cfg'
+
+def ini_get(key, cfg_file):
+    pattern = r'[ |\t]*' + key + r'[ |\t]*=[ |\t]*'
+    value_m = r'"*([^"\|\r]*)"*'
+    value = ''
+    with open(cfg_file, 'r') as ini_file:
+        for line in ini_file:
+            if re.match(pattern, line):
+                value = re.sub(pattern + value_m + '.*\n', r'\1', line)
+                break
+    return value
+
+def get_btn_num(btn, cfg):
+    num = ini_get('input_' + btn + '_btn', cfg)
+    if num: return num
+    num = ini_get('input_player1_' + btn + '_btn', cfg)
+    if num: return num
+    return ''
+
+def sysdev_get(key, sysdev_path):
+    value = ''
+    for line in open(sysdev_path + key, 'r'):
+        value = line.rstrip('\n')
+        break
+    return value
+
+def get_button_codes(dev_path):
+    js_cfg_dir = CONFIG_DIR + 'all/retroarch-joypads/'
+    js_cfg = ''
+    dev_name = ''
+    dev_button_codes = list(default_button_codes)
+
+    for device in Context().list_devices(DEVNAME=dev_path):
+        sysdev_path = os.path.normpath('/sys' + device.get('DEVPATH')) + '/'
+        if not os.path.isfile(sysdev_path + 'name'):
+            sysdev_path = os.path.normpath(sysdev_path + '/../') + '/'
+        # getting joystick name
+        dev_name = sysdev_get('name', sysdev_path)
+        # getting joystick vendor ID
+        dev_vendor_id = int(sysdev_get('id/vendor', sysdev_path), 16)
+        # getting joystick product ID
+        dev_product_id = int(sysdev_get('id/product', sysdev_path), 16)
+    if not dev_name:
+        return dev_button_codes
+
+    # getting retroarch config file for joystick
+    for f in os.listdir(js_cfg_dir):
+        if f.endswith('.cfg'):
+            input_device = ini_get('input_device', js_cfg_dir + f)
+            input_vendor_id = ini_get('input_vendor_id', js_cfg_dir + f)
+            input_product_id = ini_get('input_product_id', js_cfg_dir + f)
+            if (input_device == dev_name and
+               (input_vendor_id  == '' or int(input_vendor_id)  == dev_vendor_id) and
+               (input_product_id == '' or int(input_product_id) == dev_product_id)):
+                js_cfg = js_cfg_dir + f
+                break
+    if not js_cfg:
+        js_cfg = RETROARCH_CFG
+
+    # getting configs for dpad, buttons A, B, X and Y
+    btn_map = [ 'left', 'right', 'up', 'down', 'a', 'b', 'x', 'y' ]
```

Comment:
Also, we only need buttons a and b.  x and y should be removed as they serve no purpose from what I can tell so far.

Reply:
I think Y [never gets defined](https://github.com/ridgekuhn/arkos/blob/338d76c3912cae9cb887af10e6989e93452847a8/12272020/root/opt/joy2key/RetroPie-Setup/scriptmodules/supplementary/runcommand/joy2key.py#L117) unless it's [explicitly passed](https://github.com/ridgekuhn/arkos/blob/338d76c3912cae9cb887af10e6989e93452847a8/12272020/root/opt/joy2key/RetroPie-Setup/scriptmodules/helpers.sh#L29) to `joy2keyStart`. If you want to cut it down to two buttons, eliminating the spacebar mapping from B [in the wrapper's default params](https://github.com/ridgekuhn/arkos/blob/338d76c3912cae9cb887af10e6989e93452847a8/12272020/root/opt/joy2key/listen.sh#L55) seems most appropriate, imo. That way we don't have to modify the upstream scripts (except for the directory paths they're looking for configs in), if that makes sense? 

I'm not sure what purpose the spacebar services in dialog/whiptail anyway; I think it's like a soft return, where it lets you make selections, but doesn't submit the form unless you've selected "OK", as opposed to hitting return, which does let you submit it without selecting "OK"?

Reply:
If possible we should make A function as return so the end user can just select that without having to scroll left or right to select.  B should function as a esc key.

Reply:
Okay, so in the wrapper the default args would be `KEY_MAPPINGS=(kcub1 kcuf1 kcuu1 kcud1 0x0a 0x1b)`. I'm not a fan of scrolling left/right either, but that may be a limitation of whiptail/dialog. I need to look more into how it works, I'm just thinking about all the times I've used those whiptail/dialog boxes over the years and know that sometimes I can just hit enter, and sometimes I can't. 🤔